Document/0x04c-Tampering-and-Reverse-Engineering.md

```diff
@@ -196,14 +197,27 @@ Even though most professional GUI-based disassemblers feature scripting faciliti
+In simple words, symbolic execution is mathematically analyzing a program without executing it. During analysis, each unknown input is represented as a mathematical variable (a symbolic value), and all the operations hence performed on these unknowns is recorded in a form of mathematical equations. These mathematical equations are called `constraints`. In the end of this analysis, a final mathematical equation is obtained, in which the variables are the inputs whose values are not known. SMT solvers are special programs which solve these equations to give possible values for these variables. To visualise this, lets use a function which takes one input and multiplies it by 279 and then subtracts the output by 30. The  mathematical equation for this operation on input will be `(x * 279 - 30)`, where `x` is the variable representing the input to the function. If the value to be multiplied is another input to this function, the equation will become `(x * y - 30)`, where `y` being the another input introduced. We can see the equation is more complex than before. Moving forward, instead of subtracting a constant 30, a value from a global variable is used. This adds another variable to the equation, and makes it `(x * y - z)`. As is the case for global variables, their value can be changed from outside this function, which may lead to different outputs whenever this function is executed. This adds to additional complexity in determining correct solution. Finally, there is an _if_ condition which checks if the value calculated is greater than 1000, and returns _success_ if true, else returns _fail_. If we want the function to always return _success_, then we require a solution for the equation `(x * y - z) > 1000`. This equation can be given as input to a SMT solver to get values for x, y and z which satisfies the equation. Internally SMT solvers use various equation solving techniques to generate solution for such equations. Some of the techniques are very advance and there discussion is beyond the scope of this book.
```

Comment:
I've made the example a bit more simple to be able to explain the same using less words. What do you think? Also I changes global by external variable to justify the possibility of having to perform dynamic analysis to obtain its value on runtime (as stated some paragraphs below). 

```suggestion
In simple words, symbolic execution is mathematically analyzing a program without executing it. During analysis, each unknown input is represented as a mathematical variable (a symbolic value), and hence all the operations performed on these variables are recorded as a tree of operations (aka. AST (abstract syntax tree), from compiler theory). These ASTs can be translated into so-called _constraints_ that will be interpreted by a SMT solver. In the end of this analysis, a final mathematical equation is obtained, in which the variables are the inputs whose values are not known. SMT solvers are special programs which solve these equations to give possible values for the input variables given a final state.

To illustrate this, imagine a function which takes one input (`x`) and multiplies it by the value of a second input (`y`). Finally, there is an _if_ condition which checks if the value calculated is greater than the value of an external variable(`z`), and returns "success" if true, else returns "fail". The equation for this operation will be `(x * y) > z`.

If we want the function to always return "success" (final state), we can tell the SMT solver to calculate the values for `x` and `y` (input variables) which satisfy the corresponding equation. As is the case for global variables, their value can be changed from outside this function, which may lead to different outputs whenever this function is executed. This adds to additional complexity in determining correct solution.

Internally SMT solvers use various equation solving techniques to generate solution for such equations. Some of the techniques are very advanced and their discussion is beyond the scope of this book.
```